tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java

```diff
@@ -0,0 +1,304 @@
+package org.tensorflow.framework.losses.impl;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.Reduction;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.ReduceSum;
+import org.tensorflow.op.core.Squeeze;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.family.TNumber;
+
+import java.util.Collections;
+
+public class LossesImpl {
+
+  /**
+   * Squeeze or expand last dimension if needed with a sampleWeights of one.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of <code>predictions</code> or <code>labels</code> if their rank differs by 1 (using
+   *       {@link #removeSqueezableDimensions}).
+   *   <li>Squeezes or expands last dim of <code>sampleWeight` if its rank differs by 1 from the new
+   *       rank of <code>predictions`. If <code>sampleWeight` is scalar, it is kept scalar./li>
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>. Each of them possibly has the last
+   *     dimension squeezed, <code>sampleWeight</code> could be extended by one dimension. If <code>sampleWeight</code>
+   *     is null, (prediction, label) is returned.
+   */
+  public static <T extends TNumber> Tuple<T> squeezeOrExpandDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions) {
+    return squeezeOrExpandDimensions(tf, labels, predictions, null);
+  }
+
+  /**
+   * Squeeze or expand last dimension if needed.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of `predictions` or `labels` if their rank differs by 1 (using *
+   *       `confusion_matrix.remove_squeezable_dimensions`). *
+   *   <li>Squeezes or expands last dim of `sampleWeight` if its rank differs by 1 from the new *
+   *       rank of `predictions`. If `sampleWeight` is scalar, it is kept scalar./li> *
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction
```

Comment:
Is "match" the right way to describe the precondition relationship between `predictions` and `labels`?

Reply:
It is definitely not the same `Shape`.  I was thinking of compatible, but that has specific meaning in `Shape.isCompatibleWIth`.  The description is saying the ranks must be equal or differ by one. I am not sure of one word that describes that.  `match` was the word used in the Python version of this method.

Reply:
Hmm, here's a suggestion:
* We could decide what we want the convention to be, in terms of squeeze-or-expand plus maybe broadcasting.
* Write this up carefully in the class javadoc for either `Loss` or `Losses`.
* Mention that documentation in the class javadoc for every other loss class.
* Also mention it in `Loss#call`.
* And be silent about it in the individual methods of `Losses`.
Perhaps?

Reply:
`match` must mean that the shapes of the input operands are capable of being molded into the relationships defined for the result of this method. Again `LossesImpl` is intended to be marked as module private (JDK 11) and only should be accessible from the losses or metrics package. It is not intended to be a general use API.

Reply:
We should probably note in the javadoc for the class that this is an internal implementation class and subject to change (and being locked off under the module system).

Reply:
Added this comment for the `LossesImpl` class

```
/**
 * These are helper methods for Losses and will be module private when
 * Java modularity is applied to TensorFlow Java.
 * These methods should not be used outside of the Loss package.
 */
```